src/devices/Board/Board.cs

```diff
@@ -0,0 +1,692 @@
+        /// <param name="usage">The current pin usage</param>
+        /// <param name="owner">The current pin owner</param>
+        /// <exception cref="InvalidOperationException">The pin is not reserved, or the owner is not correct</exception>
+        public virtual void ReleasePin(int pinNumber, PinUsage usage, object owner)
+        {
+            if (!_initialized)
+            {
+                throw new InvalidOperationException("Cannot release a pin if board is not initialized.");
+            }
+
+            int logicalPin = ConvertPinNumber(pinNumber, DefaultPinNumberingScheme, PinNumberingScheme.Logical);
+
+            lock (_pinReservationsLock)
+            {
+                if (_pinReservations.TryGetValue(logicalPin, out List<PinReservation>? reservations))
+                {
+                    if (!UsageCanSharePins(usage))
+                    {
+                        PinReservation reservation = reservations.Single();
+                        if (reservation.Owner != owner || reservation.Usage != usage)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not the owner or the usage is wrong. Class {reservation.Owner} has reserved the Pin for {reservation.Usage}");
+                        }
+
+                        _pinReservations.Remove(logicalPin);
+                    }
+                    else
+                    {
+                        PinReservation? reservation = reservations.FirstOrDefault(x => x.Owner == owner);
+                        if (reservation == null)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not a valid owner.");
+                        }
+
+                        if (reservation.Usage != usage)
+                        {
+                            throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because you are not the owner or the usage is wrong. Class {reservation.Owner} has reserved the Pin for {reservation.Usage}");
+                        }
+
+                        reservations.Remove(reservation);
+                        if (reservations.Count == 0)
+                        {
+                            _pinReservations.Remove(logicalPin);
+                        }
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because it is not reserved.");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the current usage of a pin
+        /// </summary>
+        /// <param name="pinNumber">Pin number in the current numbering scheme</param>
+        /// <returns>The current usage of a pin</returns>
+        public abstract PinUsage DetermineCurrentPinUsage(int pinNumber);
```

Comment:
Why is this abstract? can't you just check the pinReservations list?

Reply:
This uses a separate list (moved it here now). The reason this is abstract (or virtual) is that an implementation is suggested to query the hardware, if possible. This allows determining the pin usage without changing it first.